sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala

```diff
@@ -53,4 +58,117 @@ class ResolveSubquerySuite extends AnalysisTest {
       JoinHint.NONE)
     assertAnalysisSuccess(expr)
   }
+
+  test("deduplicate lateral subquery") {
+    val plan = LateralJoin(t1, LateralSubquery(t0.select('a)), Inner, None)
+    // The subquery's output OuterReference(a#0) conflicts with the left child output
+    // attribute a#0. So an alias should be added to deduplicate the subquery's outputs.
+    val expected = LateralJoin(
+      t1,
+      LateralSubquery(Project(Seq(OuterReference(a).as(a.name)), t0), Seq(a)),
+      Inner,
+      None)
+    checkAnalysis(plan, expected)
+  }
+
+  test("lateral join with ambiguous join conditions") {
+    val plan = LateralJoin(t1, LateralSubquery(t0.select('b)), Inner, Some('b ===  1))
+    assertAnalysisError(plan, "Reference 'b' is ambiguous, could be: b, b." :: Nil)
+  }
+
+  test("prefer resolving lateral subquery attributes from the inner query") {
+    val plan = LateralJoin(t1, LateralSubquery(t2.select('a, 'b, 'c)), Inner, None)
+    val expected = LateralJoin(
+      t1,
+      LateralSubquery(Project(Seq(OuterReference(a).as(a.name), b, c), t2), Seq(a)),
+      Inner, None)
+    checkAnalysis(plan, expected)
+  }
+
+  test("qualified column names in lateral subquery") {
+    val t1b = b.withQualifier(Seq("t1"))
+    val t2b = b.withQualifier(Seq("t2"))
+    checkAnalysis(
+      LateralJoin(t1.as("t1"), LateralSubquery(t0.select($"t1.b")), Inner, None),
+      LateralJoin(
+        t1,
+        LateralSubquery(Project(Seq(OuterReference(t1b).as(b.name)), t0), Seq(t1b)),
+        Inner, None)
+    )
+    checkAnalysis(
+      LateralJoin(
+        t1.as("t1"),
+        LateralSubquery(t2.as("t2").select($"t1.b", $"t2.b")),
+        Inner, None),
+      LateralJoin(
+        t1,
+        LateralSubquery(Project(Seq(OuterReference(t1b).as(b.name), t2b), t2.as("t2")), Seq(t1b)),
+        Inner, None)
+    )
+  }
+
+  test("resolve nested lateral subqueries") {
+    // SELECT * FROM t1, LATERAL (SELECT * FROM t2, LATERAL (SELECT b, c))
+    val plan = LateralJoin(t1,
+      LateralSubquery(
+        LateralJoin(t2, LateralSubquery(t0.select('b, 'c)), Inner, None)),
+      Inner, None)
+    val expected = LateralJoin(t1,
+      LateralSubquery(LateralJoin(t2,
+          LateralSubquery(
+            Project(Seq(OuterReference(b).as(b.name), OuterReference(c).as(c.name)), t0),
+            Seq(b, c)),
+          Inner, None)),
+      Inner, None)
+    checkAnalysis(plan, expected)
+  }
+
+  test("lateral subquery with unresolvable attributes") {
+    // SELECT * FROM t1, LATERAL (SELECT a, c)
+    assertAnalysisError(
+      LateralJoin(t1, LateralSubquery(t0.select('a, 'c)), Inner, None),
+      Seq("cannot resolve 'c' given input columns: []")
```

Comment:
note: we can improve the error message in the future, to include columns from the outer plan as well.